### PR TITLE
#478 Speed up loading of menas

### PR DESCRIPTION
**Pull request by lokm01**
- Added a profile "generateComponentPreload" which generates
  a preload file from components
- Added a default cache time for static resources in spring
- Added library preload for sap.ui.table and sap.ui.layout
- Additional speed up can be achieved by enabling HTTP
  compression on the server.
  See KB16333: How to enable HTTP compression under Apache Tomcat
  in MicroStrategy Web Universal
https://community.microstrategy.com/s/article/KB16333-How-to-enable-HTTP-compression-under-Apache-Tomcat-in?language=en_US

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -62,3 +62,6 @@ myTestCheckpoints
 *.config
 *.keytab
 .DS_Store
+
+package-lock.json
```
Comment (Zejnilovic) [inline .gitignore]: In Spline, we have `package-lock.json` as part of the project. For more refer to [spline issue 237](https://github.com/AbsaOSS/spline/issues/237). There is even a nice [article](https://medium.com/coinmonks/everything-you-wanted-to-know-about-package-lock-json-b81911aa8ab8) that explains why to commit `package-lock.json`

Comment (Zejnilovic) [reply .gitignore]: Depending on the action here we should handle preload.js in the #453 as well

Comment (lokm01) [reply .gitignore]: Fair enough

File: pom.xml [hunk 2]
```diff
@@ -285,5 +285,8 @@
                 <module>selenium</module>
             </modules>
         </profile>
+        <profile>
```
[the remark this answers is not in the document]
Comment (lokm01) [reply pom.xml]: Yes, works without :)

File: menas/pom.xml [hunk 3]
```diff
@@ -299,4 +299,46 @@
             </testResource>
         </testResources>
     </build>
+    <profiles>
+        <profile>
+            <id>generateComponentPreload</id>
```
[the remark this answers is not in the document]
Comment (lokm01) [reply menas/pom.xml]: Created issue #486

File: README.md [hunk 4]
```diff
@@ -50,6 +50,7 @@ Ensure the properties there fit your environment.
 - Without tests: `mvn clean package -DskipTests `
 - With unit tests: `mvn clean package`
 - With integration tests: `mvn clean package -Pintegration`
+- With component preload file generated (requires npm):  `mvn clean package -PgenerateComponentPreload`
```
Comment (Zejnilovic) [inline README.md]: Please add NPM to [build requirements](https://github.com/AbsaOSS/enceladus#build-requirements) as optional requirement

Comment (GeorgiChochov) [reply README.md]: Also needs grunt.

Comment (Zejnilovic) [reply README.md]: That is actually an issue. Grunt should be pulled automatically by the build process. Jan si looking into it.

Comment (GeorgiChochov) [reply README.md]: There is a [frontend-maven-plugin](https://github.com/eirslett/frontend-maven-plugin) which can automatically install these for you as part of the maven build process. Not sure if it's something you're interested in. Just putting it out there.

Comment (lokm01) [reply README.md]: Yeah used that in the latest update. Seems to work smooth.

So in the latest version, no dependencies on npm or node, all installed automatically with the profile.

**Comment by Zejnilovic** [comment]
I get this weird output when I run it now. But the whole build passes

```bash
[INFO] Including org.scala-lang.modules:scala-xml_2.11:jar:1.0.6 in the shaded jar.
[INFO] Including za.co.absa:atum:jar:0.2.2 in the shaded jar.
[INFO] Including org.json4s:json4s-ext_2.11:jar:3.2.11 in the shaded jar.
[INFO] Including org.joda:joda-convert:jar:1.6 in the shaded jar.
[WARNING] npm WARN enceladus-menas@1.0.0 No repository field.
[WARNING] npm WARN enceladus-menas@1.0.0 license should be a valid SPDX license expression
[ERROR] 
[INFO] + grunt-openui5@0.15.0
[INFO] + grunt@0.4.5
[INFO] + grunt-cli@1.3.2
[INFO] added 315 packages from 268 contributors and audited 1271 packages in 10.499s
[INFO] found 24 vulnerabilities (7 low, 13 moderate, 4 high)
[INFO]   run `npm audit fix` to fix them, or `npm audit` for details
[INFO] 
[INFO] --- frontend-maven-plugin:1.7.6:grunt (generate preload) @ enceladus-menas ---
[INFO] Running 'grunt openui5_preload' in /Users/absz189/Development/enceladus/menas/ui
[INFO] /Users/XXX/Development/enceladus/conformance/src/main/scala:-1: info: compiling
```

**Comment by lokm01** [comment]
> I get this weird output when I run it now. But the whole build passes
> 
> ```shell
> [INFO] Including org.scala-lang.modules:scala-xml_2.11:jar:1.0.6 in the shaded jar.
> [INFO] Including za.co.absa:atum:jar:0.2.2 in the shaded jar.
> [INFO] Including org.json4s:json4s-ext_2.11:jar:3.2.11 in the shaded jar.
> [INFO] Including org.joda:joda-convert:jar:1.6 in the shaded jar.
> [WARNING] npm WARN enceladus-menas@1.0.0 No repository field.
> [WARNING] npm WARN enceladus-menas@1.0.0 license should be a valid SPDX license expression
> [ERROR] 
> [INFO] + grunt-openui5@0.15.0
> [INFO] + grunt@0.4.5
> [INFO] + grunt-cli@1.3.2
> [INFO] added 315 packages from 268 contributors and audited 1271 packages in 10.499s
> [INFO] found 24 vulnerabilities (7 low, 13 moderate, 4 high)
> [INFO]   run `npm audit fix` to fix them, or `npm audit` for details
> [INFO] 
> [INFO] --- frontend-maven-plugin:1.7.6:grunt (generate preload) @ enceladus-menas ---
> [INFO] Running 'grunt openui5_preload' in /Users/absz189/Development/enceladus/menas/ui
> [INFO] /Users/XXX/Development/enceladus/conformance/src/main/scala:-1: info: compiling
> ```

Did it generate component preload file? Should be in menas/ui/components

**Comment by Zejnilovic** [comment]
> > I get this weird output when I run it now. But the whole build passes
> > ```shell
> > [INFO] Including org.scala-lang.modules:scala-xml_2.11:jar:1.0.6 in the shaded jar.
> > [INFO] Including za.co.absa:atum:jar:0.2.2 in the shaded jar.
> > [INFO] Including org.json4s:json4s-ext_2.11:jar:3.2.11 in the shaded jar.
> > [INFO] Including org.joda:joda-convert:jar:1.6 in the shaded jar.
> > [WARNING] npm WARN enceladus-menas@1.0.0 No repository field.
> > [WARNING] npm WARN enceladus-menas@1.0.0 license should be a valid SPDX license expression
> > [ERROR] 
> > [INFO] + grunt-openui5@0.15.0
> > [INFO] + grunt@0.4.5
> > [INFO] + grunt-cli@1.3.2
> > [INFO] added 315 packages from 268 contributors and audited 1271 packages in 10.499s
> > [INFO] found 24 vulnerabilities (7 low, 13 moderate, 4 high)
> > [INFO]   run `npm audit fix` to fix them, or `npm audit` for details
> > [INFO] 
> > [INFO] --- frontend-maven-plugin:1.7.6:grunt (generate preload) @ enceladus-menas ---
> > [INFO] Running 'grunt openui5_preload' in /Users/absz189/Development/enceladus/menas/ui
> > [INFO] /Users/XXX/Development/enceladus/conformance/src/main/scala:-1: info: compiling
> > ```
> 
> Did it generate component preload file? Should be in menas/ui/components

Yes, it's there. When I said it passes, I meant correctly. But the error is still there

**Comment by lokm01** [comment]
> > > I get this weird output when I run it now. But the whole build passes
> > > ```shell
> > > [INFO] Including org.scala-lang.modules:scala-xml_2.11:jar:1.0.6 in the shaded jar.
> > > [INFO] Including za.co.absa:atum:jar:0.2.2 in the shaded jar.
> > > [INFO] Including org.json4s:json4s-ext_2.11:jar:3.2.11 in the shaded jar.
> > > [INFO] Including org.joda:joda-convert:jar:1.6 in the shaded jar.
> > > [WARNING] npm WARN enceladus-menas@1.0.0 No repository field.
> > > [WARNING] npm WARN enceladus-menas@1.0.0 license should be a valid SPDX license expression
> > > [ERROR] 
> > > [INFO] + grunt-openui5@0.15.0
> > > [INFO] + grunt@0.4.5
> > > [INFO] + grunt-cli@1.3.2
> > > [INFO] added 315 packages from 268 contributors and audited 1271 packages in 10.499s
> > > [INFO] found 24 vulnerabilities (7 low, 13 moderate, 4 high)
> > > [INFO]   run `npm audit fix` to fix them, or `npm audit` for details
> > > [INFO] 
> > > [INFO] --- frontend-maven-plugin:1.7.6:grunt (generate preload) @ enceladus-menas ---
> > > [INFO] Running 'grunt openui5_preload' in /Users/absz189/Development/enceladus/menas/ui
> > > [INFO] /Users/XXX/Development/enceladus/conformance/src/main/scala:-1: info: compiling
> > > ```
> > 
> > 
> > Did it generate component preload file? Should be in menas/ui/components
> 
> Yes, it's there. When I said it passes, I meant correctly. But the error is still there

Seems that this plugin has issues with wrong log levels from npm see https://github.com/eirslett/frontend-maven-plugin/issues/733

**Pull request merged**